clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRebalanceListener.java

```diff
@@ -122,4 +122,30 @@
      * @throws org.apache.kafka.common.errors.InterruptException If raised from a nested call to {@link KafkaConsumer}
      */
     void onPartitionsAssigned(Collection<TopicPartition> partitions);
+
+    /**
+     * A callback method the user can implement to provide handling of cleaning up resources for partitions that have already
+     * been re-assigned to other consumers. This method will not be called during normal execution as the owned partitions would
+     * first be revoked by calling the {@link ConsumerRebalanceListener#onPartitionsRevoked} first, before being re-assigned
+     * to other consumers. However, when the consumer is being kicked out of the group and hence were not aware when the new
+     * group is formed without itself, this function will then be called when the consumer finally be notified about the
+     * partitions that have already been emigrated to other consumers.
+     *
+     * By default it will just trigger {@link ConsumerRebalanceListener#onPartitionsRevoked}; for advanced users who want to distinguish
+     * the handling logic of revoked partitions v.s. emigrated partitions, they can override the default implementation.
+     *
+     * It is possible
+     * for a {@link org.apache.kafka.common.errors.WakeupException} or {@link org.apache.kafka.common.errors.InterruptException}
+     * to be raised from one these nested invocations. In this case, the exception will be propagated to the current
```

Comment:
s/one these/one of these

Reply:
This PR is on top of some open PRs from @guozhangwang -- only changes from this diff are in StickyAssignor and CooperativeStickyAssignor (will rebase after merging [part 3](https://github.com/apache/kafka/pull/6884/))